src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -546,4 +617,197 @@ private CompletableFuture<List<V>> setToValueCache(List<V> assembledValues, List
+            valuesFuture.complete(completedValues);
+        }
+
+        @Override
+        public void onError(Throwable ex) {
+            assertState(!onCompleteCalled, () -> "onComplete has already been called; onError may not be invoked.");
+            onErrorCalled = true;
+
+            stats.incrementBatchLoadExceptionCount(new IncrementBatchLoadExceptionCountStatisticsContext<>(keys, callContexts));
+            if (ex instanceof CompletionException) {
+                ex = ex.getCause();
+            }
+            // Set the remaining keys to the exception.
+            for (int i = idx; i < queuedFutures.size(); i++) {
+                K key = keys.get(i);
+                CompletableFuture<V> future = queuedFutures.get(i);
+                future.completeExceptionally(ex);
+                // clear any cached view of this key because they all failed
+                dataLoader.clear(key);
+            }
+        }
+    }
+
+    private class DataLoaderMapEntrySubscriber implements Subscriber<Map.Entry<K, V>> {
+        private final CompletableFuture<List<V>> valuesFuture;
+        private final List<K> keys;
+        private final List<Object> callContexts;
+        private final List<CompletableFuture<V>> queuedFutures;
+        private final Map<K, Object> callContextByKey;
+        private final Map<K, CompletableFuture<V>> queuedFutureByKey;
+
+        private final List<K> clearCacheKeys = new ArrayList<>();
+        private final Map<K, V> completedValuesByKey = new HashMap<>();
+        private boolean onErrorCalled = false;
+        private boolean onCompleteCalled = false;
+
+        private DataLoaderMapEntrySubscriber(
+            CompletableFuture<List<V>> valuesFuture,
+            List<K> keys,
+            List<Object> callContexts,
+            List<CompletableFuture<V>> queuedFutures
+        ) {
+            this.valuesFuture = valuesFuture;
+            this.keys = keys;
+            this.callContexts = callContexts;
+            this.queuedFutures = queuedFutures;
+
+            this.callContextByKey = new HashMap<>();
+            this.queuedFutureByKey = new HashMap<>();
+            for (int idx = 0; idx < queuedFutures.size(); idx++) {
+                K key = keys.get(idx);
+                Object callContext = callContexts.get(idx);
+                CompletableFuture<V> queuedFuture = queuedFutures.get(idx);
+                callContextByKey.put(key, callContext);
+                queuedFutureByKey.put(key, queuedFuture);
+            }
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            subscription.request(keys.size());
+        }
+
+        @Override
+        public void onNext(Map.Entry<K, V> entry) {
+            assertState(!onErrorCalled, () -> "onError has already been called; onNext may not be invoked.");
+            assertState(!onCompleteCalled, () -> "onComplete has already been called; onNext may not be invoked.");
```

Comment:
its ok to just track "finished" whether that be via onError or onComplete - you cant give a an error message as good as these but it saves a boolean in memory terms and simplifies the class